cmd/osbuild-service-maintenance/aws.go

```diff
@@ -84,33 +126,212 @@ func AWSCleanup(maxConcurrentRequests int, dryRun bool, accessKeyID, accessKey s
+	var ChildToParentIDs []ChildToParentAssociation
+
+	for _, res := range reservations {
+		for _, i := range res.Instances {
+			for _, t := range i.Tags {
+				if *t.Key == "parent" {
+					ChildToParentIDs = append(ChildToParentIDs, ChildToParentAssociation{
+						Child:  *i.InstanceId,
+						Parent: *t.Value,
+					})
+				}
+			}
+		}
+	}
+	return ChildToParentIDs
+}
+
+func checkValidParent(childId string, parent []ec2types.Reservation) bool {
+	if len(parent) == 0 {
+		logrus.Infof("Instance %s has no parent, removing it", childId)
+		return false
+	}
+	if len(parent) != 1 {
+		logrus.Errorf("Instance %s has %d parents. That should never happen, not changing anything here.", childId, len(parent))
+		return true
+	}
+	if len(parent[0].Instances) == 0 {
+		logrus.Infof("Instance %s has no parent instance, removing it", childId)
+		return false
+	}
+	if len(parent[0].Instances) != 1 {
+		logrus.Errorf("Instance %s has %d parent instances. That should never happen, not changing anything here.", childId, len(parent[0].Instances))
+		return true
+	}
+
+	parentState := parent[0].Instances[0].State.Name
+	if parentState != ec2types.InstanceStateNameTerminated {
+		return true
+	}
+	logrus.Infof("Instance %s has a parent (%s) in state %s, so we'll terminate %s.", childId, *parent[0].Instances[0].InstanceId, parentState, childId)
+	return false
+}
+
+func searchSGAndCleanup(ctx context.Context, a *awscloud.AWS, dryRun bool) error {
+	securityGroups, err := a.DescribeSecurityGroupsByPrefix(ctx, "SG for i-")
+	if err != nil {
+		return err
+	}
+
+	for _, sg := range securityGroups {
+		if sg.GroupId == nil || sg.GroupName == nil {
+			logrus.Errorf(
+				"Security Group needs to have a GroupId (%v) and a GroupName (%v).",
+				sg.GroupId,
+				sg.GroupName)
+			continue
+		}
+		reservations, err := a.DescribeInstancesBySecurityGroupID(*sg.GroupId)
+		if err != nil {
+			logrus.Errorf("Failed to describe security group %s: %v", *sg.GroupId, err)
+			continue
+		}
+
+		// If no instance is running/pending, delete the SG
+		if allTerminated(reservations) {
+			logrus.Infof("Deleting security group: %s (%s)", *sg.GroupName, *sg.GroupId)
+			if !dryRun {
+				err := a.DeleteSecurityGroupById(ctx, sg.GroupId)
+
+				if err != nil {
+					logrus.Errorf("Failed to delete security group %s: %v", *sg.GroupId, err)
+				}
+			}
+		} else {
+			logrus.Debugf("Security group %s has non terminated instances associated with it.", *sg.GroupId)
+		}
+	}
+	return nil
+}
+
+// allTerminated returns true if any instance of the reservations is not terminated
+// then it's considered "in use"
+func allTerminated(reservations []ec2types.Reservation) bool {
+	allAreTerminated := true
+
+	for _, reservation := range reservations {
+		for _, instance := range reservation.Instances {
+			if instance.State != nil && (instance.State.Name != ec2types.InstanceStateNameTerminated) {
+				allAreTerminated = false
```

Comment:
This could just `return false` here, this would avoid the need for the `allAreTerminated` var (and the last return of the function would just be `return true`)

Reply:
Thanks! …refactoring mess I left there :-/